engine/src/flutter/lib/web_ui/lib/src/engine/compositing/canvas_provider.dart

```diff
@@ -0,0 +1,123 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+/// Manages the lifecycle of raw canvas elements, abstracting away the differences
+/// between onscreen and offscreen canvases.
+///
+/// This class is responsible for:
+/// - Acquiring and releasing canvas elements.
+/// - Resizing canvases.
+/// - Attaching `webglcontextlost` event listeners and notifying the consumer.
+abstract class CanvasProvider<C extends DomEventTarget> {
+  final Map<C, ui.VoidCallback> _contextLostCallbacks = <C, ui.VoidCallback>{};
+  final Map<C, DomEventListener> _eventListeners = <C, DomEventListener>{};
+
+  /// Acquires a canvas element of a given `size`.
+  ///
+  /// The `onContextLost` callback will be invoked when the underlying rendering
+  /// context for this canvas is lost.
+  C acquireCanvas(BitmapSize size, {required ui.VoidCallback onContextLost}) {
+    final C canvas = _createCanvas(size);
+    final DomEventListener eventListener = createDomEventListener((DomEvent event) {
+      final ui.VoidCallback? callback = _contextLostCallbacks[canvas];
+      if (callback != null) {
+        callback();
+      }
+      // The canvas is no longer usable.
+      releaseCanvas(canvas);
+    });
+
+    _eventListeners[canvas] = eventListener;
+    canvas.addEventListener('webglcontextlost', eventListener);
+    _contextLostCallbacks[canvas] = onContextLost;
+    return canvas;
+  }
+
+  /// Resizes the `canvas` element to the new `size`.
+  ///
+  /// This method is responsible for updating the canvas element's dimensions
+  /// and any associated properties (e.g., CSS styles for onscreen canvases).
+  void resizeCanvas(C canvas, BitmapSize size);
+
+  /// Releases a `canvas` element, allowing it to be pooled or disposed of.
+  void releaseCanvas(C canvas) {
+    final DomEventListener? listener = _eventListeners.remove(canvas);
+    if (listener != null) {
+      canvas.removeEventListener('webglcontextlost', listener);
+    }
+    _contextLostCallbacks.remove(canvas);
+    detachCanvas(canvas);
+  }
+
+  /// Disposes of all canvases managed by this provider.
+  void dispose() {
+    List<C>.from(_contextLostCallbacks.keys).forEach(releaseCanvas);
```

Comment:
If you agree with removing the `_contextLostCallbacks` map, then this can become:
```suggestion
    List<C>.from(_eventListeners.keys).forEach(releaseCanvas);
```